internal/controller/apiextensions/composite/extra_resources.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+package composite
+
+import (
+	"context"
+	"reflect"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	kunstructured "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1beta1"
+)
+
+// MaxRequirementsIterations is the maximum number of times a Function should be
+// called, limiting the number of times it can request for extra resources,
+// capped for safety.
+const MaxRequirementsIterations = 5
+
+// A FetchingFunctionRunner wraps an underlying FunctionRunner, adding support
+// for fetching any extra resources requested by the function it runs.
+type FetchingFunctionRunner struct {
+	wrapped   FunctionRunner
+	resources ExtraResourcesFetcher
+}
+
+// NewFetchingFunctionRunner returns a FunctionRunner that supports fetching
+// extra resources.
+func NewFetchingFunctionRunner(r FunctionRunner, f ExtraResourcesFetcher) *FetchingFunctionRunner {
+	return &FetchingFunctionRunner{wrapped: r, resources: f}
+}
+
+// RunFunction runs a function, repeatedly fetching any extra resources it asks
+// for. The function may be run up to MaxRequirementsIterations times.
+func (c *FetchingFunctionRunner) RunFunction(ctx context.Context, name string, req *v1beta1.RunFunctionRequest) (*v1beta1.RunFunctionResponse, error) {
```

Comment:
This is almost exactly the `runFunction` method from d31a117147684cacd2d400b57804b44e237b287e. I tried to make this move more obvious using separate git commits but I ended up getting them all mixed up and just deciding to squash them. 😬